extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeParsingHelper.java

```diff
@@ -257,49 +270,84 @@ public static String extractVideoIdFromMixId(final String playlistId) throws Par
+                final String initialData = Parser.matchGroup1(
+                        "var\\s*ytInitialData\\s*=\\s*(\\{.*?\\});", html);
                 return JsonParser.object().from(initialData);
             }
-        } catch (JsonParserException | Parser.RegexException e) {
+        } catch (final JsonParserException | Parser.RegexException e) {
             throw new ParsingException("Could not get ytInitialData", e);
         }
     }
 
-    public static boolean isHardcodedClientVersionValid() throws IOException, ExtractionException {
-        final String url = "https://www.youtube.com/results?search_query=test&pbj=1";
+    public static boolean areHardcodedClientVersionAndKeyValid() throws IOException,
+            ExtractionException {
+        if (areHardcodedClientVersionAndKeyValidValue != null) {
+            return areHardcodedClientVersionAndKeyValidValue;
+        }
+        // @formatter:off
+        final byte[] body = JsonWriter.string()
+            .object()
+                .object("context")
+                    .object("client")
+                        .value("hl", "en")
+                        .value("gl", "GB")
+                        .value("clientName", "1")
+                        .value("clientVersion", HARDCODED_CLIENT_VERSION)
+                    .end()
+                .end()
+            .end().done().getBytes(UTF_8);
+        // @formatter:on
 
         final Map<String, List<String>> headers = new HashMap<>();
         headers.put("X-YouTube-Client-Name", Collections.singletonList("1"));
-        headers.put("X-YouTube-Client-Version", Collections.singletonList(HARDCODED_CLIENT_VERSION));
-        final String response = getDownloader().get(url, headers).responseBody();
+        headers.put("X-YouTube-Client-Version",
+                Collections.singletonList(HARDCODED_CLIENT_VERSION));
+
+        // This endpoint is fetched by the YouTube website to get the items of its main menu and is
+        // pretty lightweight (around 30kB)
+        final Response response = getDownloader().post(YOUTUBEI_V1_URL + "guide?key="
+                        + HARDCODED_KEY, headers, body);
+        final String responseBody = response.responseBody();
+        final int responseCode = response.responseCode();
 
-        return response.length() > 50; // ensure to have a valid response
+        areHardcodedClientVersionAndKeyValidValue = responseBody.length() > 5000
+                && responseCode == 200; // Ensure to have a valid response
+        return areHardcodedClientVersionAndKeyValidValue;
     }
 
     private static void extractClientVersionAndKey() throws IOException, ExtractionException {
-        final String url = "https://www.youtube.com/results?search_query=test";
-        final String html = getDownloader().get(url).responseBody();
+        // Don't extract the client version and the innertube key if it has been already extracted
+        if (!keyAndVersionExtracted) return;
```

Comment:
Maybe I'm  not understanding something, but shouldn't the function `return` if the key and version were already extracted, and not the opposite?

Reply:
Yes, with the previous change that you suggested and that I applied, it no longer makes sense. I just changed that.